### DISCUSSION TITLE
how to show pyplots in nicegui?

### DISCUSSION THREAD

**Discussion by mickey3721** [zauberzeug/nicegui#1054] [Q&A]
### Question

for example code below:
```py
import numpy as np
from matplotlib import pyplot as plt

x = np.linspace(0.0, 5.0)
y1 = np.cos(2 * np.pi * x) * np.exp(-x)
y2 = np.sin(2 * np.pi * x) * np.exp(-x)
fig, (ax1, ax2) = plt.subplots(2, 1, sharex='all')
ax1.plot(x, y1, '-')
ax1.set_title('Subplot 1')
ax2.plot(x, y2)
ax2.set_title('Subplot 2')

plt.show()
```
How to show this plot with subplots through nicegui?
Official API ui.pyplot only show a simple pyplot figure, when in subplots case how to do?
Thanks!

**Comment by falkoschindler** [comment]
Hi @mickey3721, that's an interesting question!

There are two ways to create pyplots, either state-based or object-oriented. [Here](https://towardsdatascience.com/plt-subplot-or-plt-subplots-understanding-state-based-vs-object-oriented-programming-in-pyplot-4ba0c7283f5d) is an article that deals with exactly this subject (haven't read it completely though).

The `ui.pyplot` element has been designed with the state-based approach in mind. Thus, you can add subplots like this:
```py
with ui.pyplot():
    x = np.linspace(0.0, 5.0)
    y1 = np.cos(2 * np.pi * x) * np.exp(-x)
    y2 = np.sin(2 * np.pi * x) * np.exp(-x)
    ax1 = plt.subplot(2, 1, 1)
    ax1.plot(x, y1, '-')
    ax1.set_title('Subplot 1')
    ax2 = plt.subplot(2, 1, 2, sharex=ax1)
    ax2.plot(x, y2)
    ax2.set_title('Subplot 2')
```

But now I thought, what is actually the problem with the object-oriented approach? Turns out, we just have to pass the `fig` object into the pyplot element:

```py
with ui.pyplot() as plot:
    x = np.linspace(0.0, 5.0)
    y1 = np.cos(2 * np.pi * x) * np.exp(-x)
    y2 = np.sin(2 * np.pi * x) * np.exp(-x)
    plot.fig, (ax1, ax2) = plt.subplots(2, 1, sharex='all')
    ax1.plot(x, y1, '-')
    ax1.set_title('Subplot 1')
    ax2.plot(x, y2)
    ax2.set_title('Subplot 2')
```

So, either way seems to work. 🙂

**Reply by mickey3721** [reply]
Yes, they work! Thank you so much!
